atomix/utils/src/main/java/io/atomix/utils/net/Address.java

```diff
@@ -140,43 +140,20 @@ public InetAddress address() {
    * @return the IP address
    */
   public InetAddress address(final boolean resolve) {
-    if (resolve) {
-      socketAddress = resolveAddress();
+    if (resolve || socketAddress.isUnresolved()) {
+      // the constructor will by default attempt to resolve the host, and will fallback to the an
+      // unresolved address if it couldn't
+      socketAddress = new InetSocketAddress(host, port);
       return socketAddress.getAddress();
     }
 
-    // Try to resolve address, if not successful return null
-    if (socketAddress.isUnresolved()) {
-      synchronized (this) {
```

Comment:
Why was `synchronized` needed before? Would it cause any problem if this is removed?

Reply:
I imagine because there could be a race condition where you force resolve from different threads, and you want to avoid forcing resolve unless really necessary? So you need to do the check (read) and resolve (write) in the same critical section. This looks to me more like an small optimization imo, as the end result is the same. But none of us wrote this, so I can't be 100% sure of the original intent.

Reply:
Ya. I think it is safe to remove it.